src/methods.rs

```diff
@@ -46,4 +52,101 @@ impl LintPass for MethodsPass {
             }
         }
     }
+
+    fn check_item(&mut self, cx: &Context, item: &Item) {
+        if let ItemImpl(_, _, _, None, _, ref items) = item.node {
+            for item in items {
+                let name = item.ident.name;
+                for &(method_name, n_args, self_kind, out_type, trait_name) in &TRAIT_METHODS {
+                    if_let_chain! {
+                        [
+                            name == method_name,
+                            let MethodImplItem(ref sig, _) = item.node,
+                            sig.decl.inputs.len() == n_args,
+                            out_type.matches(&sig.decl.output),
+                            self_kind.matches(&sig.explicit_self.node)
+                        ], {
+                            span_lint(cx, SHOULD_IMPLEMENT_TRAIT, item.span, &format!(
+                                "defining a method called `{}` on this type; consider implementing \
+                                 the `{}` trait", name, trait_name));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+const TRAIT_METHODS: [(&'static str, usize, SelfKind, OutType, &'static str); 30] = [
+    ("add",        2, ValueSelf,  AnyType,  "std::ops::Add"),
+    ("sub",        2, ValueSelf,  AnyType,  "std::ops::Sub"),
+    ("mul",        2, ValueSelf,  AnyType,  "std::ops::Mul"),
+    ("div",        2, ValueSelf,  AnyType,  "std::ops::Div"),
+    ("rem",        2, ValueSelf,  AnyType,  "std::ops::Rem"),
+    ("shl",        2, ValueSelf,  AnyType,  "std::ops::Shl"),
+    ("shr",        2, ValueSelf,  AnyType,  "std::ops::Shr"),
+    ("bitand",     2, ValueSelf,  AnyType,  "std::ops::BitAnd"),
+    ("bitor",      2, ValueSelf,  AnyType,  "std::ops::BitOr"),
+    ("bitxor",     2, ValueSelf,  AnyType,  "std::ops::BitXor"),
+    ("neg",        1, ValueSelf,  AnyType,  "std::ops::Neg"),
+    ("not",        1, ValueSelf,  AnyType,  "std::ops::Not"),
+    ("drop",       1, RefMutSelf, UnitType, "std::ops::Drop"),
+    ("index",      2, RefSelf,    RefType,  "std::ops::Index"),
+    ("index_mut",  2, RefMutSelf, RefType,  "std::ops::IndexMut"),
+    ("deref",      1, RefSelf,    RefType,  "std::ops::Deref"),
+    ("deref_mut",  1, RefMutSelf, RefType,  "std::ops::DerefMut"),
+    ("clone",      1, RefSelf,    AnyType,  "std::clone::Clone"),
+    ("borrow",     1, RefSelf,    RefType,  "std::borrow::Borrow"),
+    ("borrow_mut", 1, RefMutSelf, RefType,  "std::borrow::BorrowMut"),
+    ("as_ref",     1, RefSelf,    RefType,  "std::convert::AsRef"),
+    ("as_mut",     1, RefMutSelf, RefType,  "std::convert::AsMut"),
+    ("eq",         2, RefSelf,    AnyType,  "std::cmp::PartialEq"),
+    ("cmp",        2, RefSelf,    AnyType,  "std::cmp::PartialOrd"),
```

Comment:
The `cmp` function is even more tricky, because correct results depend on the return type:

* if it's `Ordering` the function should be within an `impl Ord for _`
* if it's `Option<Ordering>`, the function should be within an `impl PartialOrd for _`
* otherwise, the function is in need of renaming to disambiguate